src/urllib3/response.py

```diff
@@ -169,10 +169,15 @@ def __init__(self) -> None:
         def decompress(self, data: bytes) -> bytes:
             if not data:
                 return b""
-            return self._obj.decompress(data)  # type: ignore[no-any-return]
+            data_parts = [self._obj.decompress(data)]
+            while self._obj.eof and self._obj.unused_data:
```

Comment:
Is there any risk of an infinite loop if there is less data here than a frame's worth? Basically, is there a case where `decompress(unused_data)` results in more unused data because there is not enough data for zstd to do anything with?

Reply:
My understanding is that in that case `eof` will be `False`

Reply:
This is confirmed with some testing:

```py
>>> compressed = zstd.compress(b'foo') + zstd.compress(b'bar')
>>> obj = zstd.ZstdDecompressor().decompressobj()

>>> obj.decompress(compressed[:10])
b'f'
>>> obj.eof
False
>>> obj.unused_data
b''

>>> obj.decompress(compressed[10:])
b'oo'
>>> obj.eof
True
>>> obj.unused_data
b'(\xb5/\xfd \x03\x19\x00\x00bar'
```

Reply:
Actually, sorry, one more question here. What if there is truncated data? Is there any case where there would repeatedly be unused data if the network connection was closed early without receiving everything? Or would it throw an exception in that case?

Reply:
If the `eof` attribute is True and there is some `unused_data`, then it will go into the next loop. At that point a new object is instantiated, and the `unused_data` from the previous one is fed for decompression. And so on and so forth.

If you are concerned about possible infinite loop, then it will not happen because the `unused_data` part between each iteration of the loop is guarantied to strictly decrease in size with each iteration.

Now, if the data is truncated, then nothing special will happen. The part of the received data that can be decompressed will be returned, and that will be it. No exception is raised in that case, at least due to the code about Zstandard decompression; I would guess that the code from HTTP handling will detect that packets are lost or something like this and react accordingly.

Not sure if that answers your question, please feel free to clarify otherwise!

Reply:
Yes, that answers my question, thanks!